### ISSUE TITLE
Add basic spelling check

### ISSUE THREAD

**Issue by mstachniuk** [sebbrudzinski/Open-LaTeX-Studio#16]

**Comment by angelabp** [comment]
maybe an idea to implement http://languagetool.org/ ?

**Comment by sebbrudzinski** [comment]
Thanks for the link @angelabp  It's worth investigating that tool.

**Comment by WhiteHsu** [comment]
@sebbrudzinski 
I'm interested in this, but does anyone has experience with https://languagetool.org/ previously? Or is there any other recommended way to implement this feature?

**Comment by sebbrudzinski** [comment]
Hello @WhiteHsu 
We have no previous experience with LanguageTool. The first step in this ticket would be to familiarize with it and investigae whether we can use and if it helps us. Of course we can use any other tool/library as well.

**Comment by WhiteHsu** [comment]
I have some fast investigation and find there're some tools mentioned in internet discussions. I list them with information I know so far:
- Suggester
	- URL: http://www.softcorporation.com/products/spellcheck/
	- Supported Languages: http://www.softcorporation.com/products/spellcheck/dictionaries/
	- The Latest Release: 17 Aug, 2013
	- License: http://www.softcorporation.com/products/spellcheck/LICENSE.TXT
	- Local spellchecking API
	- Suggester provides free version for basic functions, but free version only uses one language dictionary. So if we would like to support multi-language, free version won't be enough.
	
- JOrtho
	- URL: http://jortho.sourceforge.net/
	- Supported Languages: http://jortho.sourceforge.net/
	- The Latest Release: Version 1.0 (2013-03-11)
	- License: GNU General Public License (GPL)
	- Local spellchecking API
	- This API seems able to consider. However, its latest release is provided in 2013, and some URLs in the webpage cannot work, so I'm afraid this library  may not maintained now.

- Jazzy
	- http://jazzy.sourceforge.net/
	- Because Jazzy has not been upated from 2005, I think we will not consider this.

- Google API
	- Some people says this's useful. However, Google does not maintain this service now. So we can remove this from consideration.
	
- JSpell
	- URL: http://www.jspell.com/
	- There's recommendation on this software. However, after running its demo, the spell check will be triggered to show dialog after clicking the checked word one by one, so it seems not a good way we would like to adopt.
![pic01](https://cloud.githubusercontent.com/assets/7831418/20640347/6f9b24f0-b416-11e6-9abc-ba9f93c34d1d.png)

- Language Tool
	- URL: https://languagetool.org/
	- Supported Language: https://languagetool.org/languages/
	- The Latest Release: 3.5 (2016-09-30)
	- License: Freely available under the LGPL 2.1 or later
	- Local spellchecking API
	- This tool seems being kept updated. The latest version will be released 2016-12-28. In addition, this doesn't provide Java API but also source codes and LaText tool: http://wiki.languagetool.org/checking-la-tex-with-languagetool
	So it seems Language Tool provides more resources to fulfill our requirement.

In summary, my idea is to try Language Tool firstly. How about your suggestions?  :-)

**Comment by sebbrudzinski** [comment]
Thanks for this investigation @WhiteHsu It seems like all we need at this point is Java API. I think giving Language Tool a shoot first is a good idea.

**Comment by WhiteHsu** [comment]
Sorry for late update because being too busy previously. I'll continue my development on this workitem in these days.

**Comment by WhiteHsu** [comment]
@sebbrudzinski 
Because I'm not very familiar with NetBeans IDE, I encounter several questions and would like to ask your suggestions:


1. In order to adopt Language Tool, we need to include its JAR library files. However, under the project file structure, I don't find a folder named "libs" or similar ones, so where do you suggest to put these JAR files under the project "OpenLaTeXStudio"?
![default](https://cloud.githubusercontent.com/assets/7831418/22028091/f924ce56-dd10-11e6-9542-00962989442b.png)


2. After copying the JAR files, the next I think we should add them into project. I've followed instructions based on some webpages as this one:
http://stackoverflow.com/questions/1975973/how-to-use-jar-files-in-netbeans

The steps are: Tool -> Libraries -> New Libraries -> Add JAR/Folder:
![default](https://cloud.githubusercontent.com/assets/7831418/22028211/72e2f15a-dd11-11e6-89d5-30c39f4e32e4.png)

However, after these steps, I cannot import the packages of Language Tool still. I guess there must be something missed among my steps. So may you help instruct the correct process to add library into project in NetBeans?


3. In addition to add library files, I guess we should also update pom.xml, correct? But why I find many pom.xml in my built project as below? Which is the one we should update?
![default](https://cloud.githubusercontent.com/assets/7831418/22028304/c5e19be0-dd11-11e6-954d-a3f4570e7e22.png)


Sorry for these questions. Please kindly suggest and feedback. Thank you very much.

**Comment by sebbrudzinski** [comment]
@WhiteHsu We use Maven, that handles the dependencies and external libraries. LanguageTool explains how to add their library to an existing Maven project here: http://wiki.languagetool.org/java-api (it should be as simple as adding a few lines in pom.xml)

**Comment by WhiteHsu** [comment]
@sebbrudzinski 
Thanks for your feedback. Now I'm working on this and would like to highlight the incorrect spelling check as below:
![default](https://cloud.githubusercontent.com/assets/7831418/22555416/9537890c-e99e-11e6-9985-3f08ddd25d7b.png)

So far I can import LanguageTool for spell checking, and the next is try to highlight words as screen above. However I have few questions and would like to have your suggestions:

(1) Currently in the text editor for the words in "Comment" and behind the "\" token, the words will be automatically highlighted in Green or Brown:
![default](https://cloud.githubusercontent.com/assets/7831418/22556695/09eda02a-e9a3-11e6-9dce-284a82be0741.png)
Because I do not find the codes of this automatic highlight, may you help suggest where I can find these "highlight" codes as reference, then I'll reuse them to highlight the incorrect spelling?

(2) Continue from Question 1: I find the "rSyntaxTextArea" object in the [EditorTopComponent.java] has the "setBackground" or some other similar methods. However, after calling this, it'll change background color of the entire editor. 
![default](https://cloud.githubusercontent.com/assets/7831418/22556939/c5fa290a-e9a3-11e6-850e-f97d80962cc0.png)
Because I would like to highlight some words of some parts in the editor only, do you suggest any way to implement this? Are there any codes can be reused or referred to?

(3) This's my 1st time do implement spelling check feature.  In addition to my proposed way to highlight spelling check results, do you recommend any other way to show this message to users?

Please kindly suggest and feedback. Thank you very much.

**Comment by WhiteHsu** [comment]
@sebbrudzinski 
I've checked in updates for this new feature. Here're some points about this feature:

- For now there's a new green icon as screen below:
![default](https://cloud.githubusercontent.com/assets/7831418/22587682/ab11cf7c-ea3d-11e6-84ba-9457197f0691.png)

- After clicking this icon, it'll execute spell check and highlight incorrect words:
![default](https://cloud.githubusercontent.com/assets/7831418/22587750/f4273b84-ea3d-11e6-99be-d82d2c9c4092.png)

- Then you can click the green icon again to clear highlights:
![default](https://cloud.githubusercontent.com/assets/7831418/22587733/e1f40190-ea3d-11e6-8f0f-170808f7f3c5.png)

- This "spell check" can be also triggered by clicking the option in the menu:
![default](https://cloud.githubusercontent.com/assets/7831418/22587776/16e3a5fe-ea3e-11e6-9a90-a860ca1cac6f.png)

Please help review and provide your valuable feedback. Any advice will be very appreciated. Thank you very much.  :-)

**Comment by WhiteHsu** [comment]
@sebbrudzinski 
Is there anything able to improve or enhance? Please kindly help review and provide your valuable feedback on my check-in. Thank you very much.  :-)

**Comment by sebbrudzinski** [comment]
@WhiteHsu sorry for delay. Thanks for the pull request, I've reviewed it now. Overall this works, but we need to improve several things in the code.

**Comment by WhiteHsu** [comment]
@sebbrudzinski 
Thanks for your review comments. I've updated codes based on your review comments. Please kindly help check again and provide your feedback. Thank you very much. :-)

**Comment by sebbrudzinski** [comment]
Fixed by @WhiteHsu  New tickets will be created to expand and improve the current spell check
